feat(manual_is_variant): Check for `.iter().any()`

qmwrygdxpzc opened the pull request:
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust-clippy/pull/15989)*

Now the linter checks for calling the `iter` or `into_iter` method
followed by calling the `all` or `any` method.

We have used these methods by mistake ourselves.  The developer was not
familiar with the `is_some_and` method.  He tried to use the `any`
method on an option.  The compiler suggested using `iter` first.  The
developer did not get to learn about the idiomatic method.

The lint `search_is_some` involves converting a closure that takes the
reference into one that takes the value.  The code is split between the
function for the lint and a shared function.  The new feature that we
are adding involves the same transformation.  To do so it needs some of
the same code that is currently in the `search_is_some` lint.  So it is
moved from the lint into the shared function.

Before this patch, line 799 in the file `clippy_utils/src/sugg.rs` had a
check.  It made the function return none in case the closure did not
dereference the argument.  However, the code for the lint
`search_is_some` would then fall back to using the original closure.
That achieved the same effect as the check not existing.  So the check
is now removed.

changelog: [`manual_is_variant`]: check for calling the `iter` or
`into_iter` method followed by calling the `all` or `any` method

changelog: [`manual_is_variant`]: skip the check when the MSRV is before
1.70

Commit:
refactor(search_is_some): Move some logic for transforming closures

The lint `search_is_some` involves converting a closure that takes the
reference into one that takes the value.  The code is split between the
function for the lint and a shared function.  The new feature that we
are adding involves the same transformation.  To do so it needs some of
the same code that is currently in the `search_is_some` lint.  So it is
moved from the lint into the shared function.

Before this patch, line 799 in the file `clippy_utils/src/sugg.rs` had a
check.  It made the function return none in case the closure did not
dereference the argument.  However, the code for the lint
`search_is_some` would then fall back to using the original closure.
That achieved the same effect as the check not existing.  So the check
is now removed.

Alexendoo commented on clippy_utils/src/sugg.rs:
```diff
@@ -774,8 +774,22 @@ pub fn deref_closure_args(cx: &LateContext<'_>, closure: &hir::Expr<'_>) -> Opti
+        }
+        if !matches!(strip_pat_refs(closure_arg.pat).kind, PatKind::Binding(..)) {
+            return None;
+        }
```
I believe this should still live outside of the `clippy_utils` method, though it's not currently used elsewhere it's specified as being generally applicable

qmwrygdxpzc replied:
With this patch, both `search_is_some` and `manual_is_variant_and` use it.

Alexendoo replied:
Yeah for those it might be fine, but may not be applicable to other future uses

cc @Jarcho on this though since I'm not 100% sure on what `deref_closure_args` is doing

Commit:
feat(manual_is_variant): Check for `.iter().any()`

Now the linter checks for calling the `iter` or `into_iter` method
followed by calling the `all` or `any` method.

We have used these methods by mistake ourselves.  The developer was not
familiar with the `is_some_and` method.  He tried to use the `any`
method on an option.  The compiler suggested using `iter` first.  The
developer did not get to learn about the idiomatic method.

changelog: [`manual_is_variant`]: check for calling the `iter` or
`into_iter` method followed by calling the `all` or `any` method

Commit:
fix(manual_is_variant): Skip old versions

changelog: [`manual_is_variant`]: skip the check when the MSRV is before
1.70

qmwrygdxpzc commented:
@ada4a This patch has taken a long time.  Is the reviewer busy or is there a problem with the patch?

profetia commented:
You might use `r? clippy` to find a new reviewer

qmwrygdxpzc commented:
It takes too long to get this patch reviewed.
